src/core/algorithms/nar/des/encoded_nar.cpp

```diff
@@ -0,0 +1,102 @@
+#include "encoded_nar.h"
+
+#include "algorithms/nar/value_range.h"
+#include "model/types/types.h"
+
+namespace algos::des {
+
+using model::NAR;
+
+size_t EncodedNAR::VectorSize() {
+    return encoded_value_ranges_.size() * EncodedValueRange().kFieldCount + 1;
+}
+
+size_t EncodedNAR::FeatureCount() {
+    return encoded_value_ranges_.size();
+}
+
+// TODO: remove code duplication here
+double& EncodedNAR::operator[](size_t index) {
+    qualities_consistent_ = false;
+    if (index == 0) {
+        return implication_sign_pos_;
+    } else {
+        index--;
+        size_t feature = index / EncodedValueRange().kFieldCount;
+        size_t feature_field = index % EncodedValueRange().kFieldCount;
+        return encoded_value_ranges_[feature][feature_field];
+    }
+}
+
+double const& EncodedNAR::operator[](size_t index) const {
+    if (index == 0) {
+        return implication_sign_pos_;
+    } else {
+        index--;
+        size_t feature = index / EncodedValueRange().kFieldCount;
+        size_t feature_field = index % EncodedValueRange().kFieldCount;
+        return encoded_value_ranges_[feature][feature_field];
+    }
+}
+
+NAR EncodedNAR::SetQualities(FeatureDomains domains, TypedRelation const* typed_relation) {
+    NAR this_decoded = Decode(domains);
+    this_decoded.SetQualities(typed_relation);
+    qualities_ = this_decoded.GetQualities();
+    qualities_consistent_ = true;
+    return this_decoded;
+}
+
+model::NARQualities const& EncodedNAR::GetQualities() const {
+    if (!qualities_consistent_) {
+        throw std::logic_error("Getting uninitialized qualities from NAR.");
+    }
+    return qualities_;
+}
+
+NAR EncodedNAR::Decode(FeatureDomains domains) const {
+    auto resulting_nar = model::NAR();
+
+    std::vector<size_t> feature_order(encoded_value_ranges_.size());
+    std::iota(std::begin(feature_order), std::end(feature_order), 0);
+    auto compare_by_permutation = [e = encoded_value_ranges_](size_t& a, size_t& b) -> bool {
+        return e[a].permutation > e[b].permutation;
+    };
+    std::sort(feature_order.begin(), feature_order.end(), compare_by_permutation);
+
+    uint implication_sign_after = implication_sign_pos_ * (encoded_value_ranges_.size() - 1);
+    uint handling_feat_num = -1;
```

Comment:
`uint`? Use `size_t`

Reply:
size of `encoded_value_ranges_` is not zero, right? Otherwise that's an integer overflow

Reply:
`implication_sign_after` is not used if `encoded_value_ranges_` size is zero because then the loop is not entered because `feature_order` is of size 0.

Reply:
> `uint`? Use `size_t`

Fixed. Could you tell me why size_t is better here?

Reply:
> > `uint`? Use `size_t`
> 
> Fixed. Could you tell me why size_t is better here?

I mean, it does it's job, but it's unconventional:
```bash
imorozko@112618560-nix:~/maga/VanyaVolgushev(add-des-release)$ git grep "uint "
src/core/algorithms/nar/des/encoded_nar.cpp:    uint implication_sign_after = implication_sign_pos_ * (encoded_value_ranges_.size() - 1);
src/core/algorithms/nar/des/encoded_nar.cpp:    uint handling_feat_num = -1;
src/core/model/types/null_type.h:        static std::atomic_uint hash = 0;
imorozko@112618560-nix:~/maga/VanyaVolgushev(add-des-release)$ 
imorozko@112618560-nix:~/maga/VanyaVolgushev(add-des-release)$ git grep "size_t " | wc -l
1372
```
There are two instances of `uint` usage and `1372` of `size_t`